File: programming-in-scala/doc/s8-s9/FileMatcher.java
import java.util.List;
import java.util.ArrayList;
import java.io.File;

/**
 * <p>ディレクトリ内でのファイル名検索を行うクラスです。
 *
 * <p>検索方法は後方一致、部分一致、正規表現の三つです。
 */
public class FileMatcher {
	private File[] filesHere = (new File(".")).listFiles();

	/** 後方一致検索 */
	public List<File> filesEnding(String query) {
		List<File> result = new ArrayList<File>();

		for (File file : filesHere) {
			if (file.getName().endsWith(query)) result.add(file);
		}
		return result;
	}

	/** 部分一致検索 */
	public List<File> filesContaining(String query) {
		List<File> result = new ArrayList<File>();

		for (File file : filesHere) {
			if (file.getName().contains(query)) result.add(file);
		}
		return result;
	}

	/** 正規表現 */
	public List<File> filesRegex(String query) {
		List<File> result = new ArrayList<File>();

		for (File file : filesHere) {
			if (file.getName().matches(query)) result.add(file);
		}
		return result;
	}
}

File: programming-in-scala/doc/s8-s9/FileMatcher.scala
package lib

import java.io.File

/**
 * <p>ディレクトリ内でのファイル名検索を行うクラスです。
 *
 * <p>検索方法は後方一致、部分一致、正規表現の三つです。
 */
object FileMatcher {
	private def filesHere = (new File(".")).listFiles

	private def filesMatching(matcher: String => Boolean) =
		for (file <- filesHere; if matcher(file.getName))
			yield file

	/** 後方一致検索 */
	def filesEnding(query: String) = filesMatching(_.endsWith(query))

	/** 部分一致検索 */
	def filesContaining(query: String) = filesMatching(_.contains(query))

	/** 正規表現 */
	def filesRegex(query: String) = filesMatching(_.matches(query))
}
